pkg/services/gitrepo/service.go

```diff
@@ -0,0 +1,128 @@
+package gitrepo
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/weaveworks/weave-gitops/pkg/git"
+	"github.com/weaveworks/weave-gitops/pkg/gitproviders"
+	"github.com/weaveworks/weave-gitops/pkg/logger"
+	"github.com/weaveworks/weave-gitops/pkg/services/automation"
+)
+
+type RepoWriter interface {
+	CreatePullRequest(ctx context.Context, info gitproviders.PullRequestInfo) error
+	WriteAndMerge(ctx context.Context, repoDir, commitMsg string, manifests []automation.AutomationManifest) error
+	CloneRepo(ctx context.Context, branch string) (func(), string, error)
+	GetDefaultBranch(ctx context.Context) (string, error)
+	CommitAndPush(ctx context.Context, commitMsg string, filters ...func(string) bool) error
+	Write(ctx context.Context, path string, content []byte) error
+	Remove(ctx context.Context, path string) error
+}
+
+type RepoWriterSvc struct {
+	URL         gitproviders.RepoURL
+	GitProvider gitproviders.GitProvider
+	GitClient   git.Git
+	Logger      logger.Logger
+}
+
+var _ RepoWriter = &RepoWriterSvc{}
+
+func NewRepoWriter(url gitproviders.RepoURL, gitProvider gitproviders.GitProvider, gitClient git.Git, logger logger.Logger) RepoWriter {
+	return &RepoWriterSvc{URL: url, GitProvider: gitProvider, GitClient: gitClient, Logger: logger}
+}
+
+func (rw *RepoWriterSvc) CreatePullRequest(ctx context.Context, info gitproviders.PullRequestInfo) error {
+	pr, err := rw.GitProvider.CreatePullRequest(ctx, rw.URL, info)
+	if err != nil {
+		return fmt.Errorf("unable to create pull request: %w", err)
+	}
+
+	rw.Logger.Println("Pull Request created: %s\n", pr.Get().WebURL)
+
+	return nil
+}
+
+func (rw *RepoWriterSvc) WriteAndMerge(ctx context.Context, repoDir, commitMsg string, manifests []automation.AutomationManifest) error {
+	for _, m := range manifests {
+		if err := rw.GitClient.Write(m.Path, m.Content); err != nil {
+			return fmt.Errorf("Failed to write manifest: %w", err)
+		}
+	}
+
+	return rw.CommitAndPush(ctx, commitMsg, func(fname string) bool {
+		for _, m := range manifests {
+			if fname == m.Path {
+				return true
+			}
+		}
+
+		return false
+	})
+}
+
+func (rw *RepoWriterSvc) CloneRepo(ctx context.Context, branch string) (func(), string, error) {
+	return CloneRepo(ctx, rw.GitClient, rw.URL, branch)
+}
+
+func (rw *RepoWriterSvc) CommitAndPush(ctx context.Context, commitMsg string, filters ...func(string) bool) error {
+	return CommitAndPush(ctx, rw.GitClient, commitMsg, rw.Logger, filters...)
+}
+
+func CommitAndPush(ctx context.Context, client git.Git, commitMsg string, logger logger.Logger, filters ...func(string) bool) error {
```

Comment:
Question: Why split this into it's own function? Is it meant to be used outside of a `RepoWriter`?

Reply:
It was already being used in `gitops/install.go` to do the initial repo setup